probe/host/controls.go

```diff
@@ -0,0 +1,66 @@
+package host
+
+import (
+	"os/exec"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/kr/pty"
+
+	"github.com/weaveworks/scope/common/xfer"
+	"github.com/weaveworks/scope/probe/controls"
+)
+
+// Control IDs used by the host integration.
+const (
+	ExecHost = "host_exec"
+)
+
+// Controls handles controls for a hosts.
+type Controls struct {
+	pipes controls.PipeClient
+}
+
+// NewControls creates new host controls.
+func NewControls(pipes controls.PipeClient) *Controls {
+	c := &Controls{pipes: pipes}
+	controls.Register(ExecHost, c.execHost)
+	return c
+}
+
+// Stop stops the host controls.
+func (*Controls) Stop() {
+	controls.Rm(ExecHost)
+}
+
+func (c *Controls) execHost(req xfer.Request) xfer.Response {
+	cmd := exec.Command(hostShellCmd[0], hostShellCmd[1:]...)
+	cmd.Env = []string{"TERM=xterm"}
+	ptyPipe, err := pty.Start(cmd)
+	if err != nil {
+		return xfer.ResponseError(err)
+	}
+
+	id, pipe, err := controls.NewPipeFromEnds(nil, ptyPipe, c.pipes, req.AppID)
+	if err != nil {
+		return xfer.ResponseError(err)
+	}
+	pipe.OnClose(func() {
+		if err := cmd.Process.Kill(); err != nil {
+			log.Errorf("Error closing host shell: %v", err)
+			return
+		}
+		log.Info("Host shell closed.")
+	})
+	go func() {
+		if err := cmd.Wait(); err != nil {
+			log.Errorf("Error waiting on host shell: %v", err)
+		}
+		ptyPipe.Close()
+		pipe.Close()
```

Comment:
Those `Closes()`s do different things (`pipe.Close()` calls `OnClose()` but doesn't close the `ptyPipe`) the pattern may be a bit odd but that's how the `Pipe` interface works (something to take into consideration for #1215)